### DISCUSSION TITLE
0.21.0

### DISCUSSION THREAD

**Discussion by matlabbe** [introlab/rtabmap#996] [Announcements]
## Standalone changes
 * RealSense2:
   *  D400 cameras: IR + Depth mode (with IR emitter disabled) is now used by default instead of RGB+Depth. The IR camera has a larger field-of-view, has a global shutter and is perfectly synchronized with depth image, so visual odometry is more accurate. 
   * The included realsense2 library in windows binaries has been [patched](https://github.com/IntelRealSense/librealsense/issues/9030) so that we can use [T265 alone like a stereo camera](http://official-rtab-map-forum.206.s1.nabble.com/Slam-using-Intel-RealSense-tracking-camera-T265-tp6333p6336.html).
 * GUI: Added multi-session localization view, related to this new 2022 paper [Multi-Session Visual SLAM for Illumination Invariant Re-Localization in Indoor Environments](https://github.com/introlab/rtabmap/tree/master/archive/2022-IlluminationInvariant)
 * Updated with Qt6 support
 * Updated with VTK 9 support
 * Added OAK-D Lite support

## Library changes
 * RTAB-Map library now uses CMake targets to make it easier to include in another cmake project.
    ```cmake
    find_package(RTABMap REQUIRED)
    target_link_libraries(MyApp rtabmap::core)
    ```
 * Updated [compilation instructions on Windows](https://github.com/introlab/rtabmap/wiki/Installation#vcpkg), now using officially [vcpkg](https://github.com/microsoft/vcpkg) for main dependencies.
 * Added option to set priors on markers (automatically transform map in world coordinates)
 * [ROS] Multi-camera visual odometry and localization support (build with [OpenGV](https://github.com/laurentkneip/opengv) dependency).
 * [Android] Added recovery feature, android 30 support, fixed various bugs
 * [iOS] Improved significantly rendering performance on latest iPhones
 * [Docker] Added 22.04 / Jammy image
 * Improved stability and many bug fixes as seen below.

**Full Changelog**: https://github.com/introlab/rtabmap/compare/0.20.16...0.21.0
 
### Pull requests
* Add Boost serialization package for libpointmatcher case by @MarvinStuede in https://github.com/introlab/rtabmap/pull/810
* List dependencies only once by @Timple in https://github.com/introlab/rtabmap/pull/812
* fix missing curl install in focal dockerfile by @benknight135 in https://github.com/introlab/rtabmap/pull/816
* Allow negative disparity in dialog for stereo matchers by @benknight135 in https://github.com/introlab/rtabmap/pull/817
* Marker priors by @matlabbe in https://github.com/introlab/rtabmap/pull/859
* fixing missing DISABLE_VTK on recent PCL versions by @matlabbe in https://github.com/introlab/rtabmap/pull/871
* Added stereo multi-camera support by @matlabbe in https://github.com/introlab/rtabmap/pull/884
* Use imported target for res_tool during cross compilation. by @windelbouwman in https://github.com/introlab/rtabmap/pull/865
* implement multi camera marker detection #898 by @matlabbe in https://github.com/introlab/rtabmap/pull/899
* Pnp multicam refactoring by @matlabbe in https://github.com/introlab/rtabmap/pull/902
* OpenCV 4.7 moved aruco, and has breaking changes by @cnSchwarzer in https://github.com/introlab/rtabmap/pull/955
* fix bigobj under MSVC, fix yaml-cpp link. by @cnSchwarzer in https://github.com/introlab/rtabmap/pull/957
* Refactoring RTABMapConfig.cmake with targets by @matlabbe in https://github.com/introlab/rtabmap/pull/904
* Qt6 compatibility by @matlabbe in https://github.com/introlab/rtabmap/pull/968

### New Contributors
* @MarvinStuede made their first contribution in https://github.com/introlab/rtabmap/pull/810
* @Timple made their first contribution in https://github.com/introlab/rtabmap/pull/812
* @benknight135 made their first contribution in https://github.com/introlab/rtabmap/pull/816
* @windelbouwman made their first contribution in https://github.com/introlab/rtabmap/pull/865
* @cnSchwarzer made their first contribution in https://github.com/introlab/rtabmap/pull/955

## Dependencies included in the Windows binaries: 
 * **Core**: OpenCV 4.7.0 with xfeatures2d and nonfree modules (GPU options enabled on cuda11.7 release), PCL 1.13.0 with VTK 9.2.0, Qt6.4.3.
 * **Drivers**: 
    * OpenNI and OpenNI2 (Kinect for XBOX 360, Xtion Live Pro, Orbbec Astra), 
    * Freenect2 ([UsbDk v1.00-22](https://github.com/daynix/UsbDk/releases/tag/v1.00-22) should be installed), 
    * Kinect for Windows SDK 2.0 (Kinect for XBOX ONE), 
    * Kinect For Azure, 
    * ~RealSense (ZR300, R200)~(check older releases), 
    * RealSense2 2.53.1 (T265, D415, D435, D435i, L515, [D400 Series visual presets](https://github.com/IntelRealSense/librealsense/wiki/D400-Series-Visual-Presets) can be used), 
    * Tara, 
    * ~Mynt Eye S SDK~(check older releases),
    * ZED SDK 3.8 (enabled on cuda11.7 release),
    * DepthAI (OAK-D, OAK-D Lite)
  * **Optimizers**: TORO, g2o, GTSAM, Ceres
  * **ICP**: libpointmatcher, CCCoreLib
  * **Reconstruction**: OctoMap

## Other
* Looking for ICP configuration file for L515 and Kinect Azure, see [release 0.20.8]( https://github.com/introlab/rtabmap/releases/tag/0.20.8).
* Windows High CPU Usage: this can be related to OpenMP, try setting environment variable `OMP_WAIT_POLICY` to `passive`.
* [SuperPoint](https://github.com/magicleap/SuperPointPretrainedNetwork) weights can be downloaded in assets below for convenience (tested with CUDA release).
* See [Installation](https://github.com/introlab/rtabmap/wiki/Installation#windows) page for additional installation instructions depending on your platform and the camera you are using (e.g., drivers).

<hr /><em>This discussion was created from the release <a href='https://github.com/introlab/rtabmap/releases/tag/0.21.0'>0.21.0</a>.</em>

**Comment by laitifranz** [comment]
Please, release the updated package macOS version. I cannot install it by myself using the source, even with all the workarounds that you linked in your installation tab. 

I'm stuck in the qt5 error, where it seems there is a malformed path error (the same showed here https://github.com/leela-zero/leela-zero/issues/2177). 
The old version of the .dmg available in the 0.20.3 release of RTABMap crashes when it is launched.
I'm using macOS Ventura 13.2.1 on an Intel Mac. 

Thank you for your time and dedication to this project!

**Reply by matlabbe** [reply]
Yes, I want to release binaries for Mac OS X, also planning to release new ROS1/ROS2 binaries (with new rtabmap_ros structure). 

My concern with Mac is that my old 2010 Macbook Pro intel cannot install latest Mac OS X, and my new Mac Mini is Apple Silicon. I am still not sure how well homebrew or macport support universal binaries (to work on both Mac intel and apple silicon). At least I remember [last year making RTAB-Map running on Mac Mini M1](https://github.com/introlab/rtabmap/wiki/Installation#source) (but without any camera drivers, because nobody was supporting apple silicon yet). Cross fingers that rebuilding will be easier now. At worst, it will be binaries without camera drivers, but at least could be used to view/edit/correct/replay databases taken from iOS app. I'll put an update in this thread when I will upload a DMG.

**Reply by laitifranz** [reply]
Really appreciate! Just having view/edit/correct/replay tools would be amazing. 

Thanks again

**Reply by matlabbe** [reply]
Uploaded DMG for MAC OS X ([RTABMap-0.21.0-Darwin-arm64.dmg](https://github.com/introlab/rtabmap/releases/download/0.21.0/RTABMap-0.21.0-Darwin-arm64.dmg)), currently only available
 for Apple Silicon Mac. If you have an intel Mac, current solution is to rebuild RTAB-Map. I [fixed](https://github.com/introlab/rtabmap/commit/6dca8ab0f366b2628cafc332ec5ef10869e120b3) the seg fault on start related to VTK OpenGL, and revised the [compilation instructions](https://github.com/introlab/rtabmap/wiki/Installation#mac-os-x-). It should not be hard to build it now.

**Reply by laitifranz** [reply]
Thank you very much! 
I tried to build from source with the new instruction but, unfortunately, the problem mentioned in my first comment persists... I would like to contribute to the release of an Intel version but I cannot figure out how to solve the issue... anyway, thank you again!

**Reply by matlabbe** [reply]
At the job, I think I can have access to an Intel Macbook pro. I'll give a try next week.

**Reply by matlabbe** [reply]
Uploaded intel version: https://github.com/introlab/rtabmap/releases/download/0.21.0/RTABMap-0.21.0-Darwin-Intel.dmg

On the Mac I built it I didn't need to sign it. If you have error about libraries not signed (see error report), report and I will try signing them.

**Reply by laitifranz** [reply]
It works!! Thank you very much, really appreciate your help
If macOS said to move to the trash, just open it by right-clicking on the RTABMap icon and clicking "Open", then it prompts out to open it anyway

**Comment by BorisLimTP** [comment]
trying to use my L515 camera in the windows client but i get a "Camera Initialisation failed" here's the logs: https://pastebin.com/ujBMS14q

**Reply by matlabbe** [reply]
Try selecting L515 camera from the main menu Detection->Select Source->RGBD Camera->L515. This will make sure that default realsense parameters are compatible with that camera.

**Reply by BorisLimTP** [reply]
i tried this as well but still the same errors + log

**Reply by matlabbe** [reply]
Which firmware version is installed on the camera?

**Comment by ArcSketch** [comment]
Would it be much trouble getting a code signature from Apple? - https://developer.apple.com/developer-id/

**Reply by matlabbe** [reply]
The [app released on App Store](https://apps.apple.com/lk/app/rtab-map-3d-lidar-scanner/id1564774365) should be already signed.

**Reply by ArcSketch** [reply]
Sorry, I wasn't being clear. I meant the Mac version, not the iOS version.

**Reply by matlabbe** [reply]
I see, the current Mac app is built using command line tools, not xcode GUI. The Intel and Arm versions are built with different computers (couldn't managed to efficiently/correctly cross-compile all dependencies on same computer). Ideally, if the app (at least the apple silicon one) could be published on Mac's App Store that would be the best. I'll try to take a look in upcoming days.

**Reply by ArcSketch** [reply]
Speaking of which, is it possible to make the iOS app available in the Mac App Store? It would be very practical to record with your iPhone but process it on a Mac. (As a temporary quick-fix I mean, if a native Mac app turns out to be too much trouble)

**Reply by ArcSketch** [reply]
Thank you for your dedication to the project

**Reply by matlabbe** [reply]
I would probably try to sign the desktop version first if it is not too long to do, to avoid potential confusion on Mac store.

**Reply by ArcSketch** [reply]
Yes, that’s probably a good call. The native desktop version would definitely be the ideal solution anyway. Thanks again.

**Reply by ArcSketch** [reply]
Hi, I'm just curious about getting the desktop version onto the Mac store. How is it going?

**Reply by matlabbe** [reply]
I am currently stuck with hundred of **ITMS-90238: Invalid Signature errors** after I upload to apple connect. The error codes I get when apple is processing my bundle:
 * ITMS-90238: Invalid Signature - The main app bundle RTAB-Map at path RTABMap.app has following signing error(s): invalid Info.plist (plist or signature have been modified) In architecture: arm64 . Refer to the Code Signing and Application Sandboxing Guide at http://developer.apple.com/library/mac/#documentation/Security/Conceptual/CodeSigningGuide/AboutCS/AboutCS.html and Technical Note 2206 at https://developer.apple.com/library/mac/technotes/tn2206/_index.html for more information.
 * ITMS-90238: Invalid Signature - The nested app bundle at path RTABMap.app/Contents/Frameworks/QtCore.framework has following signing error(s): valid on disk /Volumes/workspace/app_data/SWValidationService/mz_5628755434060812228dir/mz_6617017911165904953dir/com.introlab.rtabmap.desktop.pkg/Payload/RTABMap.app/Contents/Frameworks/QtCore.framework/Versions/5: satisfies its Designated Requirement test-requirement: code failed to satisfy specified code requirement(s) . Refer to the Code Signing and Application Sandboxing Guide at http://developer.apple.com/library/mac/#documentation/Security/Conceptual/CodeSigningGuide/AboutCS/AboutCS.html and Technical Note 2206 at https://developer.apple.com/library/mac/technotes/tn2206/_index.html for more information. 
* ITMS-90238  [... 337 more times]
 * ITMS-90296: App sandbox not enabled - The following executables must include the 'com.apple.security.app-sandbox' entitlement with a Boolean value of true in the entitlements property list: [[RTABMap.app/Contents/MacOS/RTABMap]] Refer to App Sandbox page at https://developer.apple.com/documentation/security/app_sandbox for more information on sandboxing your app. 
 * Warnings
   * ITMS-90889: 'Cannot be used with TestFlight because the bundle at “RTABMap.app” is missing a provisioning profile. Main bundles are expected to have provisioning profiles in order to be eligible for TestFlight.'

    * ITMS-90788: Incomplete Document Type Configuration - The CFBundleDocumentTypes dictionary array in the 'com.introlab.rtabmap.desktop' Info.plist should contain an LSHandlerRank value for the CFBundleTypeName 'RTAB-Map Database' entry. Refer to https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/TP40009249-SW1 for more information on the LSHandlerRank key. 

My current signing process:
```
codesign --deep --force -s "Apple Development: FIRSTNAME LASTNAME (XXXXXXXXX)" RTABMap.app
codesign --force -s "Apple Development: FIRSTNAME LASTNAME (XXXXXXXXX)" --entitlements RTABMap.app/Contents/RTABMap.entitlements  RTABMap.app/Contents/MacOS/RTABMap
xcrun productbuild --component RTABMap.app /Applications/ rtabmap.unsigned.pkg
xcrun productsign --sign "3rd Party Mac Developer Installer: FIRSTNAME LASTNAME (XXXXXXXXX)" rtabmap.unsigned.pkg rtabmap.pkg
xcrun altool --validate-app --file rtabmap.pkg --type osx 
xcrun altool --upload-app -f rtabmap.pkg -t osx 
```

I try to follow this [thread](https://developer.apple.com/forums/thread/128166) but it seems there are a lot of issues when we don't build/distribute code with XCode. It is quite difficult to find help on command line tools to generate/manage the right config files and correctly sign all dependencies.

My test above is only for Arm64 build. My main concern is how to publish universal binaries (arm64+intel) without Xcode, I would need to merge the two DMG into a single one. It doesn't seem to be just copying *.app from both architectures in same DMG. I saw that `lipo` command line app could be used to create universal binaries, but a script would be needed to merge arm64 and intel *.app together (lipo should be called on any exe in the bundle, though not sure it works with embedded libraries/plugins/frameworks). Well, the workaround would be to only distribute intel version on MacOS store, and let rosetta running it on apple silicon (I feel it not a good solution as most new mac users are now using apple silicon I guess).

I am more focused on Linux/ROS/ROS2 right now, so it is difficult for me to justify putting more time on this as the DMGs are already available from Github (https://github.com/introlab/rtabmap/releases/tag/0.21.0)

**Reply by ArcSketch** [reply]
That list really looks like a nightmare, especially all those signature errors. I completely understand that it doesn't feel like the right thing to prioritize right now. Thanks for looking into it anyway, I appreciate it.
